Foundation/NSString.swift

```diff
@@ -231,7 +231,13 @@ open class NSString : NSObject, NSCopying, NSMutableCopying, NSSecureCoding, NSC
     }
     
     open func copy(with zone: NSZone? = nil) -> Any {
-        return self
+        if let contents = _fastContents,
+            type(of: self) === NSString.self || type(of: self) === NSMutableString.self {
+            return NSString(characters: contents, length: length)
+        }
+        let result = NSString()
+        result._storage = self._storage
+        return result
     }
```

Comment:
This doesn't make a lot of sense to me. Why shouldn't an immutable `NSString` be allowed to return self for a copy?

Reply:
@parkera you're absolutely right, I didn't think about it. I updated the commit adding the case of the immutable NSString copy, what do you think about it?

Reply:
even though this method will not do what the name of it says, it makes sense to return self for performance reasons

Reply:
I think the best answer here is going to depend on if we can get a factory pattern proposal through swift-evolution... the way it works in the Darwin Foundation today is that when you use any of NSString's init methods, you're most likely getting a subclass of NSString, which implements copy by returning self. 

For now, I think we should do this:

1. if `type(of: self) === NSString.self` then return self
2. else, use one of NSString's other public initializers to create a new instance with the contents.
3. Don't assume `result._storage` exists, because a subclass would not have access to that and may not use it to hold the contents